go/fix-chain/log.go

```diff
@@ -0,0 +1,236 @@
+package fix_chain
+
+import (
+	"bytes"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/google/certificate-transparency/go/x509"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+func DumpChainPEM(chain []*x509.Certificate) string {
+	var p string
+	for _, cert := range chain {
+		b := pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}
+		p += string(pem.EncodeToMemory(&b))
+	}
+	return p
+}
+
+func DumpPEM(cert []byte) string {
+	b := pem.Block{Type: "CERTIFICATE", Bytes: cert}
+	return string(pem.EncodeToMemory(&b))
+}
+
+type toLog struct {
+	chain   []*x509.Certificate
+	retries uint16
+}
+
+type Log struct {
+	url    string
+	roots  *x509.CertPool
+	posts  chan *toLog
+	active uint32
+	// these counters are not atomically updated, so may not be quite right
+	posted        int
+	reposted      int
+	latereposted  int
+	chainreposted int
+
+	wg             sync.WaitGroup // Note that this counts the
+				      // number of active requests,
+				      // not active servers, because
+				      // we can't close it to signal
+				      // the end, because of retries.
+	postCache      map[[HashSize]byte]bool
+	postChainCache map[[HashSize]byte]bool
+	pcMutex        sync.Mutex
+	errors         chan *FixError
+}
+
+func (s *Log) Posted(cert *x509.Certificate) bool {
+	return s.postCache[Hash(cert)]
+}
+
+func (s *Log) Roots() *x509.CertPool {
+	if s.roots == nil {
+		s.roots = s.getRoots()
+	}
+	return s.roots
+}
+
+func (s *Log) getRoots() *x509.CertPool {
+	rootsjson, err := http.Get(s.url + "/ct/v1/get-roots")
+	if err != nil {
+		log.Fatalf("can't get roots from %s: %s", s.url, err)
+	}
+	defer rootsjson.Body.Close()
+	if rootsjson.StatusCode != 200 {
+		log.Fatalf("can't deal with status other than 200: %d",
+			rootsjson.StatusCode)
+	}
+	j, err := ioutil.ReadAll(rootsjson.Body)
+	//log.Printf("roots: %s", j)
+	type Certificates struct {
+		Certificates [][]byte
+	}
+	var certs Certificates
+	err = json.Unmarshal(j, &certs)
+	if err != nil {
+		log.Fatalf("can't parse json (%s): %s", err, j)
+	}
+	//log.Printf("certs: %#v", certs)
+	ret := x509.NewCertPool()
+	for i := 0; i < len(certs.Certificates); i++ {
+		r, err := x509.ParseCertificate(certs.Certificates[i])
+		switch err.(type) {
+		case nil:
```

Comment:
I think you could have this as:
`case nil, x509.NonFatalErrors:`
And maybe add a `// ignore` comment?